[stable/rabbitmq-ha] Add custom definitions

kaizensparc opened the pull request:
Since v1.9.0, a definitions.json file have been created, needed for users
creation for health check, prometheus, etc.
It is thus impossible to get these features using a custom definitions.json
This commit allows you to add :
 - Users
 - Vhosts
 - Permissions
 - Policies (already present, just moved)
 - Queues
 - Exchanges
 - Bindings

<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

Chart maintainer: @etiennetremel 

**Which issue this PR fixes**: fixes #7553

kaizensparc committed:
[stable/rabbitmq-ha] Add custom definitions

Since v1.9.0, a definitions.json file have been created, needed for users
creation for health check, prometheus, etc.
It is thus impossible to get these features using a custom definitions.json
This commit allows you to add :
 - Users
 - Vhosts
 - Permissions
 - Policies (already present, just moved)
 - Queues
 - Exchanges
 - Bindings

Signed-off-by: Clement Decoodt <clement@decoodt.eu>

steven-sheehy commented on stable/rabbitmq-ha/README.md:
```diff
@@ -67,7 +67,14 @@ and their default values.
+| `definitions.queues`                           | Pre-created queues | `{}` |
+| `definitions.exchanges`                        | Pre-created exchanges | `{}` |
+| `definitions.bindings`                         | Pre-created bindings | `{}` |
+| `definitions.policies`                         | HA policies to add to definitions.json | `{}` |
```
Changing the path of policies breaks existing users, so chart version should see a minor version increase to 1.10.0.

kaizensparc replied:
Would you prefer to keep the policies path as before?

steven-sheehy replied:
I don't have a strong opinion either way. One could argue the fact that they're stored in a definitions.json is an implementation detail and not relevant for the helm chart user. You may still need a minor version increment anyway due to new features being added, but lets let helm chart maintainers weigh in on that.

kaizensparc replied:
Ok. I keep it like this, waiting for further discussion to know if I increase version to `1.10.0` or if I put back the policies element in the root values. RFC.

steven-sheehy replied:
rabbitmq-ha chart 1.10.0 is now released, so you'll need to bump to fix conflict. I think you can keep as definitions.* but bump chart to 1.11.0.

kaizensparc replied:
I just bumped to `1.11.0`

steven-sheehy commented on stable/rabbitmq-ha/values.yaml:
```diff
@@ -13,20 +13,66 @@ managementPassword: E9R3fjZm4ejFkVFE
+#   }
+  vhosts: |-
+#   {
+#     "name": "/"
```
There's already a vhost and permission for / present in the definitions. It would be more accurate and less confusing to show a different vhost name being added as an example.

kaizensparc replied:
I made a `/rabbit` vhost. Hope it is not ambiguous.

steven-sheehy commented on stable/rabbitmq-ha/README.md:
```diff
@@ -67,7 +67,14 @@ and their default values.
 | `existingSecret`                               | Use an existing secret for password & erlang cookie                                                                                                                                                   | `""`                                                       |
-| `extraConfig`                                  | Additional configuration to add to default configmap                                                                                                                                                  | ``                                                         |
+| `extraConfig`                                  | Additional configuration to add to default configmap                                                                                                                                                  | `{}`                                                         |
+| `definitions.users`                            | Additional users | `{}` |
```
The default values for all these should be empty string since the values are actually a string containing json, not a yaml map.

kaizensparc replied:
Done, it's now 
```
`""`
```

steven-sheehy commented on stable/rabbitmq-ha/templates/_helpers.tpl:
```diff
@@ -79,10 +83,21 @@ users, virtual hosts, permissions and parameters) to load by the management plug
+    }{{- if .Values.definitions.permissions -}},
+{{ .Values.definitions.permissions | indent 4 }}
+{{- end }}
+  ]{{- if .Values.definitions.policies -}},
```
I think it's cleaner the way it was before to have an empty policies array if `.Values.policies` is empty versus having the if statements. I don't think it causes any issues in rabbitmq to list all the keys whether their contents are empty or not.

kaizensparc replied:
Should I do the same as what did `policies` before with the other fields or let it like that? It does not cause any issues (I added these statements to get a nicely readable rendered file, at the price of a not-that-nicely readable template :))

steven-sheehy replied:
I think it's still nicely rendered even if it's an empty array, so the old approach would be preferred. E.g.
```json
"policies" : [],
"queues" : [],
...
```

kaizensparc replied:
Ok. Are there any other items I should add in the definitions?

steven-sheehy replied:
I think there's also a parameters section that can be added to the definitions. Please double check

kaizensparc replied:
I added the parameters section support. I cannot find any schema or template of a full `definitions.json` file on the internet, so I cannot figure out if there are others (in the examples I found, the only lacking item was the `parameters`). Hope it's ok.

Now the default rendered file looks like this:
```json
{
  "users": [
    {
      "name": "management",
      "password": "E9R3fjZm4ejFkVFE",
      "tags": "management"
    },
    {
      "name": "guest",
      "password": "4LJFVHDVnY2G9252D1gb5VO2",
      "tags": "administrator"
    }
  ],
  "vhosts": [
    {
      "name": "/"
    }
  ],
  "permissions": [
    {
      "user": "guest",
      "vhost": "/",
      "configure": ".*",
      "read": ".*",
      "write": ".*"
    }
  ],
  "parameters": [

  ],
  "policies": [

  ],
  "queues": [

  ],
  "exchanges": [

  ],
  "bindings": [

  ]
}
```

kaizensparc committed:
[stable/rabbitmq-ha]: Add parameters in definitions

Fix default values in readme.
Simplify the definitions.json template, and let all the items being
rendered (with empty arrays if empty)

Signed-off-by: Clement Decoodt <clement@decoodt.eu>

etiennetremel reviewed (approved):
LGTM, thanks!
